### PR TITLE
Add access monitoring rules services and backend

### PR DESCRIPTION
**Pull request by EdwardDowling**
Add local backend and service for Access Monitoring Rules

Updates [3132](https://github.com/gravitational/teleport.e/issues/3132)

### CODE REVIEW

File: api/types/accessmonitoringrule/convert/v1/access_monitoring_rule.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+// ToProto converts an internal access monitoring rule into a v1 access monitoring rule object.
+func ToProto(amr *accessmonitoringrule.AccessMonitoringRule) *accessmonitoringrulev1.AccessMonitoringRule {
+	subjects := make([]string, 0, len(amr.Spec.Subjects))
+	subjects = append(subjects, amr.Spec.Subjects...)
+	states := make([]string, 0, len(amr.Spec.States))
+	states = append(states, amr.Spec.States...)
+	recipients := make([]string, 0, len(amr.Spec.Notification.Recipients))
+	recipients = append(recipients, amr.Spec.Notification.Recipients...)
```
Comment (capnspacehook) [inline api/types/accessmonitoringrule/convert/v1/access_monitoring_rule.go:72]: Is the goal to copy these slices? If so using the `slices` package would be easier and more readable

File: lib/services/access_monitoring_rules.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+	"github.com/gravitational/teleport/api/client/accessmonitoringrules"
+	"github.com/gravitational/teleport/api/types/accessmonitoringrule"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+var _ AccessMonitoringRules = (*accessmonitoringrules.Client)(nil)
+
+// AccessMonitoringRules is the AccessMonitoringRule service
+type AccessMonitoringRules interface {
+	CreateAccessMonitoringRule(ctx context.Context, in *accessmonitoringrule.AccessMonitoringRule) (*accessmonitoringrule.AccessMonitoringRule, error)
```
Comment (capnspacehook) [inline lib/services/access_monitoring_rules.go:35]: Do we really need Create, Update and Upsert API methods?

File: lib/services/simple/access_monitoring_rules.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	accessMonitoringRulePrefix      = "access_monitoring_rule"
+	accessMonitoringRuleMaxPageSize = 100
+)
+
+// AccessMonitoringRuleService is a simple access monitoring rules backend service for use specifically by the cache.
```
Comment (capnspacehook) [inline lib/services/simple/access_monitoring_rules.go:38]: This service seems basically identical to the one defined in `lib/service/local`, except that there are no calls to `CheckAndSetDefaults` here. Is that omission worth creating an other service just for the cache?

**Review by capnspacehook — COMMENTED** [review]
Unless I missed it I don't see a gRPC server implementation for this service, unless it was added in an earlier PR or will be coming in a future one?

File: api/types/accessmonitoringrule/access_monitoring_rule.go [hunk 4]
```diff
@@ -16,10 +16,136 @@ limitations under the License.
...
+// AccessMonitoringRule represents a AccessMonitoringRule instance
+type AccessMonitoringRule struct {
+	// Metadata is the rules's metadata.
+	Metadata header.Metadata
+	// Kind is a resource kind
+	Kind string
+	// SubKind is an optional resource sub kind, used in some resources
+	SubKind string
+	// Version is the resource version
+	Version string
```
Comment (smallinsky) [inline api/types/accessmonitoringrule/access_monitoring_rule.go:46]: Missing `json` and  `yaml` tags ?

File: api/types/accessmonitoringrule/access_monitoring_rule.go [hunk 5]
```diff
@@ -16,10 +16,136 @@ limitations under the License.
...
+// NewAccessMonitoringRule creates a new AccessMonitoringRule resource.
+func NewAccessMonitoringRule(metadata header.Metadata, spec Spec) *AccessMonitoringRule {
+	return &AccessMonitoringRule{
+		Metadata: metadata,
+		Spec:     spec,
+	}
+}
+
+// CheckAndSetDefaults checks validity of all parameters and sets defaults.
+func (p *AccessMonitoringRule) CheckAndSetDefaults() error {
```
[the remark this answers is not in the document]
Comment (EdwardDowling) [reply api/types/accessmonitoringrule/access_monitoring_rule.go:81]: Do you mean in a similar way to "local/databaseobjectimportrole.go"?

**Review by rosstimothy — COMMENTED** [review]
Can we add some test coverage for both the RPC service and the backend service?

File: lib/services/access_monitoring_rules.go (hunk 2 above)
Comment (capnspacehook) [reply lib/services/access_monitoring_rules.go:35]: Yes but Upsert may not be required to be exposed to the gRPC service is specifically what I'm saying

Comment (capnspacehook) [reply lib/services/access_monitoring_rules.go:35]: Oh missed that you mentioned that, my bad

File: api/types/accessmonitoringrule/access_monitoring_rule.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (EdwardDowling) [reply api/types/accessmonitoringrule/access_monitoring_rule.go]: Is something like this what you were thinking?

File: api/types/accessmonitoringrule/access_monitoring_rule.go [hunk 6]
```diff
@@ -16,10 +16,82 @@ limitations under the License.
...
 // AccessMonitoringRuleSubkind represents the type of the AccessMonitoringRule, e.g., access request, MDM etc.
 type AccessMonitoringRuleSubkind string
 
 const (
 	// AccessMonitoringRuleSubkindUnknown is returned when no AccessMonitoringRule subkind matches.
 	AccessMonitoringRuleSubkindUnknown AccessMonitoringRuleSubkind = ""
 )
+
+// AccessMonitoringRule represents a AccessMonitoringRule instance
+type AccessMonitoringRule struct {
```
[the remark this answers is not in the document]
Comment (EdwardDowling) [reply api/types/accessmonitoringrule/access_monitoring_rule.go:38]: That would make more sense, I'll go change this and the follow up Prs

**Review by smallinsky — APPROVED** [review]

**Review by rosstimothy — COMMENTED** [review]
Mostly LGTM. Just waiting for the proto conversions to the pure go struct to be removed

File: lib/accessmonitoringrules/accessmonitoringrulesv1/service.go [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+// ServiceConfig holds configuration options for the access monitoring rules service.
+type ServiceConfig struct {
+	Backend    services.AccessMonitoringRules
+	Authorizer authz.Authorizer
+	Cache      Cache
+}
+
+// Cache is the subset of the cached resources that the service queries.
+type Cache interface {
+	services.AccessMonitoringRules
```
[the remark this answers is not in the document]
Comment (EdwardDowling) [reply lib/accessmonitoringrules/accessmonitoringrulesv1/service.go]: Have swapped to using just get/list and removed the embeded service.AccessMonitoringRules

File: lib/accessmonitoringrules/accessmonitoringrulesv1/service.go [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+		return nil, trace.Wrap(err)
+	}
+	if err := authCtx.CheckAccessToKind(types.KindAccessMonitoringRule, types.VerbDelete); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if err := s.backend.DeleteAllAccessMonitoringRules(ctx); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &emptypb.Empty{}, nil
+}
```
[the remark this answers is not in the document]
Comment (EdwardDowling) [reply lib/accessmonitoringrules/accessmonitoringrulesv1/service.go]: Have removed this method

File: api/types/accessmonitoringrule/access_monitoring_rule.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (EdwardDowling) [reply api/types/accessmonitoringrule/access_monitoring_rule.go]: Ah my mistake messed up a commit, they should be gone now

**Review by rosstimothy — APPROVED** [review]

**Pull request merged**